lib/charms/layer/docker_registry.py

```diff
@@ -0,0 +1,307 @@
+    else:
+        if charm_config.get('http-host'):
+            netloc = urlparse(charm_config['http-host']).netloc
+        else:
+            netloc = '"{}:{}"'.format(hookenv.unit_private_ip(),
+                                      charm_config['registry-port'])
+        insecure_registry = netloc
+
+    templating.render('daemon.json', '/etc/docker/daemon.json',
+                      {'registries': insecure_registry})
+    host.service_restart('docker')
+
+
+def get_tls_sans(relation_name=None):
+    '''Get all sans for our TLS certificate.
+
+    Return all IP/DNS data that should included as alt names when we request
+    a TLS cert. This includes our public/private address, local DNS name, any
+    configured hostname, and the address of any related proxy.
+
+    :return: sorted list of sans
+    '''
+    charm_config = hookenv.config()
+    sans = [
+        hookenv.unit_private_ip(),
+        hookenv.unit_public_ip(),
+        socket.gethostname(),
+    ]
+    if charm_config.get('http-host'):
+        http_host = urlparse(charm_config['http-host']).hostname
+        sans.append(http_host)
+
+    if relation_name:
+        proxy_sans = [hookenv.ingress_address(rid=u.rid, unit=u.unit)
+                      for u in hookenv.iter_units_for_relation_name(relation_name)]
+        sans.extend(proxy_sans)
+
+    return sorted(sans)
+
+
+def start_registry(name=None, run_args=None):
+    '''Start a registry container.
+
+    If the named registry container doesn't exist, create and start a new
+    container. Subsequent calls will start the existing named container. If a
+    name is not specified, this method will use the configured 'registry-name'
+    value.
+
+    :param name: Name of the container to start
+    :param run_args: list of additional args to pass to docker run
+    '''
+    charm_config = hookenv.config()
+    image = charm_config.get('registry-image')
+    port = charm_config.get('registry-port')
+    if not name:
+        name = charm_config.get('registry-name')
+
+    cmd = ['docker', 'container', 'list', '--all',
+           '--filter', 'name={}'.format(name), '--format', '{{.Names}}']
+    try:
+        containers = subprocess.check_output(cmd).decode('utf8')
+    except subprocess.CalledProcessError as e:
+        hookenv.log('Could not list existing containers: {}'.format(e),
+                    level=hookenv.WARNING)
+        containers = ''
+
+    if name in containers:
+        # start existing container
+        cmd = ['docker', 'container', 'start', name]
+        try:
+            subprocess.check_call(cmd)
+        except subprocess.CalledProcessError:
+            hookenv.log('Could not start existing container: {}'.format(name),
+                        level=hookenv.ERROR)
+            raise
+    else:
+        # NB: config determines the port, but the container always listens to 5000 internally
+        cmd = ['docker', 'run', '-d', '-p', '{}:5000'.format(port), '--restart', 'unless-stopped']
```

Comment:
Do you think we should have done a systemd config?

Reply:
I don't think so.  The `--restart` flag will ensure the container is started when the docker service comes up.